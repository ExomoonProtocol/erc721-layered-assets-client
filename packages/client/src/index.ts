export * from "./models";
export * from "./lib";
export * from "./utils";
