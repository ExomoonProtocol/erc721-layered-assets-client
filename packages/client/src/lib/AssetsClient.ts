/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable no-unused-vars */
/* eslint-disable @typescript-eslint/no-unused-vars */
import {
  CollectionInfo,
  AssetsObject,
  Trait,
  ConditionalRenderingConfig,
} from "../models";
import axios from "axios";
import { ModelsUtils } from "../utils/ModelsUtils";

/**
 * Parameters for initializing a new AssetsClient instance.
 */
export interface AssetsClientInitParams {
  /**
   * Base URL to be used for fetching assets (N.B. the asets should be served as static files, and should be compliant with the Exomoon ERC721 Layered NFT standard)
   * It can be a link to a public S3 bucket or a web server you own.
   * @example "https://example.com"
   * @example "http://localhost:3000"
   **/
  baseUrl: string;

  /**
   * Specifies if client should use cache for fetching assets.
   * True by default.
   * @default true
   */
  useCache?: boolean;
}

/**
 * Base fetching parameters
 * (empty for now, will be used for future extensions)
 */
export interface BaseFetchingParams {}

/**
 * Assets client class. Represents a client for fetching assets.
 * It implements methods for fetching all the necessary assets about a Exomoon ERC721 Layered NFT collection: collection info, traits, variations, colors.
 * It also provides a caching mechanism for the fetched assets, in order to avoid fetching the same assets multiple times.
 *
 */
export class AssetsClient {
  private _baseUrl: string;

  private _useCache: boolean;

  private _cachedAssetsInfo: AssetsObject;

  constructor(initParams: AssetsClientInitParams) {
    this._baseUrl = initParams.baseUrl;
    this._useCache = initParams.useCache || false;

    this.initCache();
  }

  protected initCache() {
    this._cachedAssetsInfo = {};
  }

  protected async fetchResource(url: string): Promise<any> {
    console.log(`[AssetsClient] Fetching resource: ${url}`);
    const res = await axios.get(url);
    return res;
  }

  /**
   * Gets the base URL set during initialization
   * @returns Base URL string
   * @public
   */
  public get baseUrl(): string {
    return this._baseUrl;
  }

  protected validateParams(params: BaseFetchingParams): BaseFetchingParams {
    return {
      ...params,
    };
  }

  /**
   * Fetches collection info from remote server. If cache is enabled, it will return the cached collection info if it exists.
   * @param params Extra fetching parameters
   * @returns Collection info object, or null if not found
   * @public
   */
  public async fetchCollectionInfo(
    params?: BaseFetchingParams
  ): Promise<CollectionInfo | null> {
    const validatedParams = this.validateParams(params || {});

    let collectionInfoObj: CollectionInfo | null = null;

    if (this._useCache && this._cachedAssetsInfo?.collectionInfo) {
      collectionInfoObj = this._cachedAssetsInfo.collectionInfo;
    } else {
      const response = await this.fetchResource(
        `${this.baseUrl}/collection.json`
      );

      const rawObj =
        ModelsUtils.instance.serializer.deserializeObject<CollectionInfo>(
          response.data,
          CollectionInfo
        );

      console.log("[AssetsClient] Collection info raw:", rawObj);

      collectionInfoObj = rawObj || null;

      if (collectionInfoObj) {
        this._cachedAssetsInfo.collectionInfo = collectionInfoObj;
      }
    }

    return collectionInfoObj;
  }

  /**
   * Fetches traits info from remote server. If cache is enabled, it will return the cached traits if they exist.
   * @param params Extra fetching parameters
   * @returns Traits list array, or empty array if not found
   * @public
   */
  public async fetchTraits(params?: BaseFetchingParams): Promise<Array<Trait>> {
    const validatedParams = this.validateParams(params || {});
    const collectionInfo = await this.fetchCollectionInfo({});

    let traits: Array<Trait> = [];

    if (
      this._useCache &&
      this._cachedAssetsInfo?.traits &&
      this._cachedAssetsInfo.traits.length == collectionInfo?.traitsOrder.length
    ) {
      traits = this._cachedAssetsInfo.traits;
    } else {
      const traitNames = collectionInfo?.traitsOrder || [];
      traits = await Promise.all(
        traitNames.map(async (traitName) => {
          return await this.fetchTrait(traitName);
        })
      );

      if (traits.length > 0) {
        this._cachedAssetsInfo.traits = traits;
      }
    }

    return traits;
  }

  /**
   * Fetches a single trait from remote server. If cache is enabled, it will return the cached trait if it exists.
   * @param traitName Trait name
   * @param params Extra fetching parameters
   * @returns Trait object
   * @public
   */
  public async fetchTrait(
    traitName: string,
    params?: BaseFetchingParams
  ): Promise<Trait> {
    const validatedParams = this.validateParams(params || {});

    const response = await this.fetchResource(
      `${this.baseUrl}/traits/${traitName}/trait.json`
    );
    const rawObj = ModelsUtils.instance.serializer.deserializeObject<Trait>(
      response.data,
      Trait
    );

    if (rawObj) {
      // Cache trait
      if (this._cachedAssetsInfo.traits) {
        const traitIndex = this._cachedAssetsInfo.traits.findIndex(
          (t) => t.name === traitName
        );
        if (traitIndex !== -1) {
          this._cachedAssetsInfo.traits[traitIndex] = rawObj;
        } else {
          this._cachedAssetsInfo.traits.push(rawObj);
        }
      }

      return rawObj;
    } else {
      throw new Error("Trait not found");
    }
  }

  /**
   * Fetches all assets info from remote server. If cache is enabled, it will return the cached assets info if it exists.
   * The `AssetsObject` contains the collection info and the traits list, so it's a single object containing all the necessary assets info: after fetching this object, you can access all the other assets info through the cached object.
   * @param params Extra fetching parameters
   * @returns Assets object
   * @public
   */
  public async fetchAssetsObject(
    _params?: BaseFetchingParams
  ): Promise<AssetsObject> {
    const collectionInfo = await this.fetchCollectionInfo();
    const traits = await this.fetchTraits();

    this._cachedAssetsInfo.collectionInfo = collectionInfo!;
    this._cachedAssetsInfo.traits = traits;

    return this._cachedAssetsInfo;
  }

  /**
   * Get cached assets object. If no assets are cached, it will return null.
   * This method should be called after being sure that the assets have been fetched, otherwise it will return null.
   *
   * @returns Cached assets object
   * @public
   */
  public getAssetsObject(): AssetsObject | null {
    return this._cachedAssetsInfo || null;
  }

  /**
   * Get traits list. If no traits are cached, it will return an empty array.
   * This method should be called after being sure that the traits have been fetched, otherwise it will return an empty array.
   *
   * @returns Traits list array
   * @public
   */
  public getTrais(): Trait[] {
    return this._cachedAssetsInfo.traits || [];
  }

  /**
   * Get a single trait by name. If the trait is not cached, it will return null.
   * This method should be called after being sure that the trait has been fetched, otherwise it will return null.
   *
   * @param traitName Trait name
   * @returns Trait object, or null if not found
   * @public
   */
  public getTrait(traitName: string): Trait | null {
    if (this._cachedAssetsInfo.traits) {
      return (
        this._cachedAssetsInfo.traits.find((t) => t.name === traitName) || null
      );
    }

    return null;
  }

  /**
   * Get a single trait by index. If the trait is not cached, it will return null.
   * This method should be called after being sure that the trait has been fetched, otherwise it will return null.
   *
   * @param index Trait index
   * @returns Trait object, or null if not found
   * @public
   */
  public getTraitByIndex(index: number): Trait | null {
    if (this._cachedAssetsInfo.traits) {
      return this._cachedAssetsInfo.traits[index] || null;
    }

    return null;
  }

  /**
   * Get collection info. If no collection info is cached, it will return null.
   * This method should be called after being sure that the collection info has been fetched, otherwise it will return null.
   *
   * @returns Collection info object, or null if not found
   * @public
   */
  public getCollectionInfo(): CollectionInfo | null {
    return this._cachedAssetsInfo.collectionInfo || null;
  }

  /**
   * Clears the cache of fetched assets.
   * After calling this method, all the fetched assets will be removed from the cache, and the next time you call a method that fetches assets, it will fetch them again from the remote server.
   * @public
   */
  public clearCache(): void {
    this.initCache();
  }

  /**
   * Evals the image url for the provided trait, variation and color (optional).
   * If a conditional trait config is provided, it will override the default trait, variation and color, according to the conditional trait config itself.
   *
   * @param params Parameters for getting the image URL
   * @returns Image URL string
   * @public
   */
  public getTraitImageUrl(params: {
    traitName: string;
    variationName: string;
    colorName?: string;
    conditionalTraitConfig?: ConditionalRenderingConfig;
  }): string {
    const trait = this.getTrait(params.traitName);

    if (!trait) {
      throw new Error(`Trait ${params.traitName} not found`);
    }

    let traitUrlSection = trait.name;
    const variationUrlSection = params.variationName;
    const colorUrlSection = params.colorName
      ? params.colorName
      : params.variationName;

    if (params.conditionalTraitConfig) {
      const traits = this.getTrais();

      let configMatchedWithTrait = false;

      const config = params.conditionalTraitConfig;
      const trait = traits.find((t) => t.name === config.traitName);

      if (trait) {
        // If there is no variation name, use match only by trait name
        if (!config.variationName) {
          configMatchedWithTrait = true;
        }

        // If there is a variation name, match by trait name and variation name
        const variation = trait.variations.find(
          (v) => v.name === config.variationName
        );

        if (variation) {
          if (config.colorName) {
            // If there is a color name, match by trait name, variation name and color name
            const color = variation.colors.find((c) => c === config.colorName);

            if (color) {
              configMatchedWithTrait = true;
            }
          } else {
            // If there is no color name, match by trait name and variation name
            configMatchedWithTrait = true;
          }
        }

        if (configMatchedWithTrait) {
          traitUrlSection = `${params.traitName}/${config.folderName}`;
        }
      }
    }

    const imageUrl = `${this.baseUrl}/traits/${traitUrlSection}/${variationUrlSection}/${colorUrlSection}.png`;

    return imageUrl;
  }
}
