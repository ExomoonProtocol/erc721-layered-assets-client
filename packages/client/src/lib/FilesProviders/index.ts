export * from "./BaseFileProvider";
export * from "./HttpFileProvider";
