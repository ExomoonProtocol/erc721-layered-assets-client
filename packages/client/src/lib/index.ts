export * from "./AssetsClient";
export * from "./ItemConfiguration";
export * from "./TraitConfiguration";
