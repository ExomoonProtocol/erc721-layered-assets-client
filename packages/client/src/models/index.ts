export * from "./CollectionInfo";
export * from "./Trait";
export * from "./Traits";
