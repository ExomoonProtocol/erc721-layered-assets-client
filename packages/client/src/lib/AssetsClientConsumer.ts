import { AssetsClient } from "./AssetsClient";

/**
 * Base class for consumers of AssetsClient
 */
export class AssetsClientConsumer {
  private _assetsClient: AssetsClient;

  constructor(client: AssetsClient) {
    this._assetsClient = client;
  }

  /**
   * Get assets client instance used by the consumer
   */
  public get assetsClient(): AssetsClient {
    return this._assetsClient;
  }
}
