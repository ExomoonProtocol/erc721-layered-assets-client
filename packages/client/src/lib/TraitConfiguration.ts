import { AssetsClient } from "./AssetsClient";
import { AssetsClientConsumer } from "./AssetsClientConsumer";

export class TraitConfiguration extends AssetsClientConsumer {
  private _traitName: string;

  private _variationName: string;

  private _colorName?: string;

  constructor(_client: AssetsClient) {
    super(_client);
  }

  public get traitName(): string {
    return this._traitName;
  }

  public set traitName(value: string) {
    this._traitName = value;
  }

  public get variationName(): string {
    return this._variationName;
  }

  public set variationName(value: string) {
    this._variationName = value;
  }

  public get colorName(): string | undefined {
    return this._colorName;
  }

  public set colorName(value: string | undefined) {
    this._colorName = value;
  }

  public get metadataTraitName(): string {
    return this._traitName;
  }

  public get metadataTraitValue(): string {
    return `${this._variationName || ""}${
      this._colorName ? ` ${this._colorName}` : ""
    }`;
  }

  // public getImageUrl(): string {
  //   // return this.assetsClient.getTraitImageUrl(
  //   //   this._traitName,
  //   //   this._variationName,
  //   //   this._colorName
  //   // );
  // }
}
