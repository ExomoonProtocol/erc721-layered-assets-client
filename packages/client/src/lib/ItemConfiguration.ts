/* eslint-disable no-unused-vars */

import { AssetsClient } from "./AssetsClient";
import { AssetsClientConsumer } from "./AssetsClientConsumer";
import { TraitConfiguration } from "./TraitConfiguration";

export enum ItemConfigurationStatus {
  Unloaded = 0,
  Ready,
}

export class ItemConfiguration extends AssetsClientConsumer {
  private _traitConfigurations: Array<TraitConfiguration>;

  private _status: ItemConfigurationStatus;

  constructor(_client: AssetsClient) {
    super(_client);

    this._status = ItemConfigurationStatus.Unloaded;
    this._traitConfigurations = new Array<TraitConfiguration>();
  }

  /**
   * Prepare item configuration, and make sure all assets are loaded.
   */
  public async load(): Promise<void> {
    this._status = ItemConfigurationStatus.Ready;
    await this.assetsClient.fetchAssetsObject();
  }

  /**
   * Get trait configuration by trait name
   */
  public get traitConfigurations(): Array<TraitConfiguration> {
    return this._traitConfigurations;
  }

  /**
   * Get status of item configuration
   */
  public get status(): ItemConfigurationStatus {
    return this._status;
  }

  /**
   * Require item configuration to be ready
   */
  protected requireReady(): void {
    if (this._status !== ItemConfigurationStatus.Ready) {
      throw new Error("Item configuration not ready");
    }
  }

  /**
   * Set variation for a trait
   * @param trait Trait name
   * @param variation Variation name
   * @param color Color name
   */
  public setVariation(trait: string, variation: string, color?: string): void {
    this.requireReady();

    const traitConfiguration = this._traitConfigurations.find(
      (tc) => tc.traitName === trait
    );

    const traitObj = this.assetsClient.getTrait(trait);

    if (!traitObj) {
      throw new Error(`Trait ${trait} not found`);
    }

    const variationObj = traitObj.variations.find((v) => v.name === variation);

    if (!variationObj) {
      throw new Error(`Variation ${variation} not found`);
    }

    const variationHasColors =
      variationObj.colors && variationObj.colors.length > 0;

    let selectedColor: string | null = null;

    if (variationHasColors) {
      if (color) {
        selectedColor = variationObj.colors.find((c) => c === color) || null;

        if (!selectedColor) {
          throw new Error(`Color ${color} not found`);
        }
      } else {
        selectedColor = variationObj.colors[0];
      }
    } else if (color) {
      throw new Error("Variation doesn't support colors");
    }

    if (traitConfiguration) {
      traitConfiguration.variationName = variation;
      traitConfiguration.colorName = selectedColor || undefined;
    } else {
      const newTraitConfiguration = new TraitConfiguration(this.assetsClient);
      newTraitConfiguration.traitName = trait;
      newTraitConfiguration.variationName = variation;
      newTraitConfiguration.colorName = selectedColor || undefined;
      this._traitConfigurations.push(newTraitConfiguration);
      this.sortTraitConfigurations();
    }
  }

  /**
   * Sort trait configurations based on the order of traits in collection info
   */
  protected sortTraitConfigurations(): void {
    this._traitConfigurations = (
      this.assetsClient.getCollectionInfo()?.traitsOrder || []
    )
      .map((traitName) => {
        console.log(traitName);
        return this._traitConfigurations.find(
          (tc) => tc.traitName === traitName
        ) as TraitConfiguration;
      })
      .filter((tc) => tc);
  }

  /**
   * Get trait image urls
   * @returns Array of trait image urls
   */
  public getTraitsUrls(): Array<string> {
    this.requireReady();

    return this._traitConfigurations.map((tc) => {
      // Getting info about trait (asset info)
      const traitAsset = this.assetsClient.getTrait(tc.traitName);
      console.log("Asset: ", traitAsset);

      // Check if there is a conditional rendering config for this trait
      if (traitAsset?.conditonalRenderingConfig) {
        const conditionalRenderingConfig =
          traitAsset.conditonalRenderingConfig.find(
            (currentConditionalConfig) => {
              const matchedConfiguration = this._traitConfigurations.find(
                (tcToMatch) => {
                  let matched = true;

                  if (
                    currentConditionalConfig.traitName !== tcToMatch.traitName
                  ) {
                    matched = false;
                  } else if (
                    currentConditionalConfig.variationName &&
                    currentConditionalConfig.variationName !==
                      tcToMatch.variationName
                  ) {
                    matched = false;
                  } else if (
                    currentConditionalConfig.colorName &&
                    currentConditionalConfig.colorName !== tcToMatch.colorName
                  ) {
                    matched = false;
                  }

                  return matched;
                }
              );

              return matchedConfiguration;
            }
          );

        if (conditionalRenderingConfig) {
          console.log(conditionalRenderingConfig);
          return tc.getImageUrl(conditionalRenderingConfig);
        }
      }

      return tc.getImageUrl();
    });
  }
}
