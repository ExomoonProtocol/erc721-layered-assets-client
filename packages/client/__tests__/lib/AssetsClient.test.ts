import { AssetsClient } from "../../src/lib/AssetsClient";
import { ConditionalRenderingConfig } from "../../src/models";
import { HttpServerMock } from "../HttpServerMock";
import path from "path";

describe("AssetsClient", () => {
  afterEach(() => {
    HttpServerMock.instance.clearTemporary();
  });

  describe("fetchCollectionInfo method", () => {
    it("Should fetch collectionInfo", async () => {
      HttpServerMock.instance.addTemporaryResponseMappings([
        {
          url: "https://example.com",
          filePath: path.resolve(__dirname, "../assets/1/"),
        },
      ]);

      const client = new AssetsClient({ baseUrl: "https://example.com" });
      const collectionInfo = await client.fetchCollectionInfo({});
      expect(collectionInfo).toMatchSnapshot("a1 - collectionInfo");
    });

    it("Should fetch collectionInfo with cache", async () => {
      HttpServerMock.instance.addTemporaryResponseMappings([
        {
          url: "https://example.com",
          filePath: path.resolve(__dirname, "../assets/1/"),
        },
      ]);

      const client = new AssetsClient({
        baseUrl: "https://example.com",
        useCache: true,
      });
      await client.fetchCollectionInfo({});
      HttpServerMock.instance.clearTemporary();
      const collectionInfo = await client.fetchCollectionInfo({});
      expect(collectionInfo).toMatchSnapshot("a1 - collectionInfo cached");
    });
  });

  describe("fetchTraits method", () => {
    it("Should fetch traits", async () => {
      HttpServerMock.instance.addTemporaryResponseMappings([
        {
          url: "https://example.com",
          filePath: path.resolve(__dirname, "../assets/1/"),
        },
      ]);

      const client = new AssetsClient({ baseUrl: "https://example.com" });
      const traits = await client.fetchTraits();
      expect(traits).toMatchSnapshot("a1 - traits");
    });

    it("Should fetch traits with cache", async () => {
      HttpServerMock.instance.addTemporaryResponseMappings([
        {
          url: "https://example.com",
          filePath: path.resolve(__dirname, "../assets/1/"),
        },
      ]);

      const client = new AssetsClient({
        baseUrl: "https://example.com",
        useCache: true,
      });
      await client.fetchTraits({});
      HttpServerMock.instance.clearTemporary();
      const traits = await client.fetchTraits({});
      expect(traits).toMatchSnapshot("a1 - traits cached");
    });
  });

  describe("fetchAssetsObject method", () => {
    it("Should fetch assets", async () => {
      HttpServerMock.instance.addTemporaryResponseMappings([
        {
          url: "https://example.com",
          filePath: path.resolve(__dirname, "../assets/1/"),
        },
      ]);

      const client = new AssetsClient({ baseUrl: "https://example.com" });
      const assets = await client.fetchAssetsObject();
      expect(assets).toMatchSnapshot("a1 - assets");
    });

    it("Should fetch assets with cache", async () => {
      HttpServerMock.instance.addTemporaryResponseMappings([
        {
          url: "https://example.com",
          filePath: path.resolve(__dirname, "../assets/1/"),
        },
      ]);

      const client = new AssetsClient({
        baseUrl: "https://example.com",
        useCache: true,
      });
      await client.fetchAssetsObject({});
      HttpServerMock.instance.clearTemporary();
      const assets = await client.fetchAssetsObject({});
      expect(assets).toMatchSnapshot("a1 - assets cached");
    });
  });

  describe("getTraitImageUrl method", () => {
    it("Should get trait image url of a normal variation with color", async () => {
      HttpServerMock.instance.addTemporaryResponseMappings([
        {
          url: "https://example.com",
          filePath: path.resolve(__dirname, "../assets/1/"),
        },
      ]);

      const client = new AssetsClient({ baseUrl: "https://example.com" });
      await client.fetchAssetsObject();
      const url = client.getTraitImageUrl({
        traitName: "Background",
        variationName: "Single Color",
        colorName: "Blue",
      });
      expect(url).toBe(
        "https://example.com/traits/Background/Single Color/Blue.png"
      );
    });

    it("Should get trait image url of a normal variation without color", async () => {
      HttpServerMock.instance.addTemporaryResponseMappings([
        {
          url: "https://example.com",
          filePath: path.resolve(__dirname, "../assets/1/"),
        },
      ]);

      const client = new AssetsClient({ baseUrl: "https://example.com" });
      await client.fetchAssetsObject();
      const url = client.getTraitImageUrl({
        traitName: "Background",
        variationName: "Custom 1",
      });
      expect(url).toBe(
        "https://example.com/traits/Background/Custom 1/Custom 1.png"
      );
    });

    it("Should get trait image url with conditional rendering config", async () => {
      HttpServerMock.instance.addTemporaryResponseMappings([
        {
          url: "https://example.com",
          filePath: path.resolve(__dirname, "../assets/2/"),
        },
      ]);

      const client = new AssetsClient({ baseUrl: "https://example.com" });
      await client.fetchAssetsObject();
      const url = client.getTraitImageUrl({
        traitName: "Cover",
        variationName: "Rectangle",
        conditionalTraitConfig: new ConditionalRenderingConfig({
          traitName: "Shape",
          variationName: "Rectangle",
        }),
      });
      expect(url).toBe(
        "https://example.com/traits/Cover/Shape_Rectangle/Rectangle/Rectangle.png"
      );
    });

    it("Should throw an error if trait not found", async () => {
      const client = new AssetsClient({ baseUrl: "https://example.com" });
      expect(() =>
        client.getTraitImageUrl({
          traitName: "Background",
          variationName: "Single Color",
          colorName: "Blue",
        })
      ).toThrow("Trait Background not found");
    });
  });

  describe("getPreviewImageUrl method", () => {
    it("Should get preview image url of a normal variation", async () => {
      HttpServerMock.instance.addTemporaryResponseMappings([
        {
          url: "https://example.com",
          filePath: path.resolve(__dirname, "../assets/1/"),
        },
      ]);

      const client = new AssetsClient({ baseUrl: "https://example.com" });
      await client.fetchAssetsObject();
      const url = client.getPreviewImageUrl({
        traitName: "Background",
        variationName: "Single Color",
      });
      expect(url).toBe(
        "https://example.com/traits/Background/Single Color/thumbnail.png"
      );
    });

    it("Should get preview image with conditional rendering config", async () => {
      HttpServerMock.instance.addTemporaryResponseMappings([
        {
          url: "https://example.com",
          filePath: path.resolve(__dirname, "../assets/2/"),
        },
      ]);

      const client = new AssetsClient({ baseUrl: "https://example.com" });
      await client.fetchAssetsObject();
      const url = client.getPreviewImageUrl({
        traitName: "Cover",
        variationName: "Rectangle",
        conditionalTraitConfig: new ConditionalRenderingConfig({
          traitName: "Shape",
          variationName: "Rectangle",
        }),
      });
      expect(url).toBe(
        "https://example.com/traits/Cover/Shape_Rectangle/Rectangle/thumbnail.png"
      );
    });
  });

  describe("getTraitsGroups method", () => {
    it("Should get traits groups", async () => {
      HttpServerMock.instance.addTemporaryResponseMappings([
        {
          url: "https://example.com",
          filePath: path.resolve(__dirname, "../assets/2/"),
        },
      ]);

      const client = new AssetsClient({ baseUrl: "https://example.com" });
      await client.fetchAssetsObject();
      const groups = client.getTraitsGroups();
      expect(groups).toMatchSnapshot();
    });
  });
});
