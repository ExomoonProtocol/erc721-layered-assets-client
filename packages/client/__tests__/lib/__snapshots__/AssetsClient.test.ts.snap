// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetsClient fetchAssetsObject method Should fetch assets with cache: a1 - assets cached 1`] = `
{
  "collectionInfo": CollectionInfo {
    "collectionName": "Collection 1",
    "traitsOrder": [
      "Background",
    ],
  },
  "traits": [
    Trait {
      "description": "A simple background layer",
      "iconUrl": "https://example.com/traits/Background/icon.svg",
      "name": "Background",
      "variations": [
        Variation {
          "colors": [
            "Black",
            "Blue",
            "Brown",
            "Green",
          ],
          "description": "A single color background",
          "name": "Single Color",
          "previewImageUrl": "https://example.com/traits/Background/thumbnail.png",
          "rarity": "common",
        },
        Variation {
          "description": "A custom background",
          "name": "Custom 1",
          "previewImageUrl": "https://example.com/traits/Background/thumbnail.png",
          "rarity": "rare",
        },
      ],
    },
  ],
}
`;

exports[`AssetsClient fetchAssetsObject method Should fetch assets: a1 - assets 1`] = `
{
  "collectionInfo": CollectionInfo {
    "collectionName": "Collection 1",
    "traitsOrder": [
      "Background",
    ],
  },
  "traits": [
    Trait {
      "description": "A simple background layer",
      "iconUrl": "https://example.com/traits/Background/icon.svg",
      "name": "Background",
      "variations": [
        Variation {
          "colors": [
            "Black",
            "Blue",
            "Brown",
            "Green",
          ],
          "description": "A single color background",
          "name": "Single Color",
          "previewImageUrl": "https://example.com/traits/Background/thumbnail.png",
          "rarity": "common",
        },
        Variation {
          "description": "A custom background",
          "name": "Custom 1",
          "previewImageUrl": "https://example.com/traits/Background/thumbnail.png",
          "rarity": "rare",
        },
      ],
    },
  ],
}
`;

exports[`AssetsClient fetchCollectionInfo method Should fetch collectionInfo with cache: a1 - collectionInfo cached 1`] = `
CollectionInfo {
  "collectionName": "Collection 1",
  "traitsOrder": [
    "Background",
  ],
}
`;

exports[`AssetsClient fetchCollectionInfo method Should fetch collectionInfo: a1 - collectionInfo 1`] = `
CollectionInfo {
  "collectionName": "Collection 1",
  "traitsOrder": [
    "Background",
  ],
}
`;

exports[`AssetsClient fetchTraits method Should fetch traits with cache: a1 - traits cached 1`] = `
[
  Trait {
    "description": "A simple background layer",
    "iconUrl": "https://example.com/traits/Background/icon.svg",
    "name": "Background",
    "variations": [
      Variation {
        "colors": [
          "Black",
          "Blue",
          "Brown",
          "Green",
        ],
        "description": "A single color background",
        "name": "Single Color",
        "previewImageUrl": "https://example.com/traits/Background/thumbnail.png",
        "rarity": "common",
      },
      Variation {
        "description": "A custom background",
        "name": "Custom 1",
        "previewImageUrl": "https://example.com/traits/Background/thumbnail.png",
        "rarity": "rare",
      },
    ],
  },
]
`;

exports[`AssetsClient fetchTraits method Should fetch traits: a1 - traits 1`] = `
[
  Trait {
    "description": "A simple background layer",
    "iconUrl": "https://example.com/traits/Background/icon.svg",
    "name": "Background",
    "variations": [
      Variation {
        "colors": [
          "Black",
          "Blue",
          "Brown",
          "Green",
        ],
        "description": "A single color background",
        "name": "Single Color",
        "previewImageUrl": "https://example.com/traits/Background/thumbnail.png",
        "rarity": "common",
      },
      Variation {
        "description": "A custom background",
        "name": "Custom 1",
        "previewImageUrl": "https://example.com/traits/Background/thumbnail.png",
        "rarity": "rare",
      },
    ],
  },
]
`;
